### PR TITLE
Bluetooth: ATT: remove BT_ATT_ENFORCE_FLOW

### LINKED ISSUE #65910: Bluetooth: Remove `BT_ATT_ENFORCE_FLOW`
ATT mandates that only one transaction be open at any given time per-bearer and per direction.
That means in practice only one indication, one request etc.

The current host implementation tries to enforce this by using the HCI num complete packets event as a signal that a response has been sent and that transaction is over.
This is not compatible with all controllers: see https://github.com/zephyrproject-rtos/zephyr/issues/64967

What is illegal and ATT could instead try to detect is the peer pipelining ATT transactions, e.g. batching write requests before having received a response. Enforcing this is debatable, as it's not our job to validate that the peer's host follows spec.

What ATT should instead do, is error out when it is not able to allocate a resource due the peer not respecting the spec.

Validating that a response has been sent (and the transaction is over) can be done at the handover to L2CAP TX.

TLDR: The host doesn't need to validate peer behavior, it only needs to protect itself.

### PR DESCRIPTION
**Pull request by jori-nordic**
Enforcing the peer's behavior is not strictly necessary. All the host
should do is make sure it is resilient to a spec-violating peer.

Moreover, a growing number of platforms were disabling the check, as the
spec allows "batching" HCI num complete packets events, stalling ATT RX.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/65910

### CODE REVIEW

**Comment by jori-nordic** [comment]
Draft until https://github.com/zephyrproject-rtos/zephyr/pull/65967 is merged.
Some changes are currently cherry-picked from it.

File: boards/posix/nrf_bsim/Kconfig.defconfig [hunk 1]
```diff
@@ -81,6 +81,26 @@ choice LOG_MODE
 	default LOG_MODE_IMMEDIATE
 endchoice
 
+# Since flash & ram is "free", we can also include the function names
+# in all logging levels.
+config LOG_FUNC_NAME_PREFIX_ERR
```
[the remark this answers is not in the document]
Comment (jori-nordic) [reply boards/posix/nrf_bsim/Kconfig.defconfig:86]: @aescolar do you think it's fine that I rather update this one test that's failing? Ie remove those default kconfigs just for that one test.

**Review by alwa-nordic — DISMISSED** [review]

**Comment by jori-nordic** [comment]
Ready for review

Ignore the first two commits, they are cherry-picked from https://github.com/zephyrproject-rtos/zephyr/pull/66083

**Review by theob-pro — DISMISSED** [review]

**Comment by alwa-nordic** [comment]
https://github.com/zephyrproject-rtos/zephyr/pull/66083 is now merged.

**Comment by jori-nordic** [comment]
> #66083 is now merged.

CI passed but you forgot a param ¯\_(ツ)_/¯
Added it as topmost commit.

**Comment by LingaoM** [comment]
> > #66083 is now merged.
> 
> CI passed but you forgot a param ¯_(ツ)_/¯ Added it as topmost commit.

It's broke CI, should tag `Hotfix` for this PR or move this commit to separate PR if this PR keep open for a while.

**Comment by jori-nordic** [comment]
This time it's regular "main is broken"
image

opened hotfix PR here https://github.com/zephyrproject-rtos/zephyr/pull/66146

**Review by theob-pro — APPROVED** [review]

**Review by alwa-nordic — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Pull request merged**